packages/ec-core/src/generator/wrappers/owned.rs

```diff
@@ -0,0 +1,57 @@
+use std::{borrow::Borrow, marker::PhantomData, num::NonZeroUsize};
+
+use rand::{distributions::Uniform, prelude::Distribution};
+
+use crate::generator::wrappers::slice_cloning::EmptySlice;
+
+#[derive(Debug, PartialEq)]
+pub struct OneOfCloning<T, U> {
+    collection: T,
+    range: Uniform<usize>,
+    _p: PhantomData<U>,
+}
+
+impl<T, U> OneOfCloning<T, U>
+where
+    T: Borrow<[U]>,
+{
+    /// Create a new [`OneOfCloning`] distribution, which selects a
+    /// value from a collection and then returns a new value by cloning the
+    /// selected value.
+    ///
+    /// # Errors
+    /// - [`EmptySlice`] if an empty collection is passed in, since then no
+    ///   element can be selected from there
+    pub fn new(val: T) -> Result<Self, EmptySlice> {
+        match val.borrow().len() {
+            0 => Err(EmptySlice),
+            len => Ok(Self {
+                collection: val,
+                range: Uniform::new(0, len),
+                _p: PhantomData,
+            }),
+        }
+    }
+
+    pub fn choices(&self) -> NonZeroUsize {
+        // Safety: at construction time, it was ensured that the slice was
+        // non-empty, as such the len is > 0.
+        unsafe { NonZeroUsize::new_unchecked(self.collection.borrow().len()) }
+    }
+}
+
+impl<'a, T, U> Distribution<U> for OneOfCloning<T, U>
+where
+    T: Borrow<[U]>,
+    U: 'a + Clone,
+{
+    fn sample<R: rand::Rng + ?Sized>(&self, rng: &mut R) -> U {
+        let idx = self.range.sample(rng);
+
+        // Safety: at construction time, it was ensured that the slice was
+        // non-empty, and that the `Uniform` range produces values in range
+        // for the slice
+        let slice = self.collection.borrow();
+        unsafe { slice.get_unchecked(idx) }.clone()
```

Comment:
Now another `unsafe`!

TBH, I'd be inclined to actually do all this with safe code first, and then see if there's actually a performance issue here worthy of the introduction of `unsafe` code. I can actually imagine that this might pay off since we're probably sampling this kind of distribution a fair bit, but I don't _know_ that, and I think I'd like to avoid introducing unsafe code unless there's a demonstrable reason for it.

Reply:
and I think all the unsafes we use here are really easy to wrap your head around (no raw pointers or something like that) so I think that should be fine. If you wish to keep the ec crate clean of unsafe code tho, we could move the distribution module to a choice_distributions crate if you want to

Reply:
I'll need to think about that some. I agree that adding an `.unwrap()` isn't pretty, but there is a part of me that would kinda like having this project free of unsafe code, though.

Reply:
I can add extensive testing to all the unsafe parts and we can run miri in ci if that helps